### ISSUE TITLE
Feature: Allow to use domains as handles, e.g. @domain.tld with webfinger acct:*@domain.tld

### ISSUE THREAD

**Issue by Ciantic** [mastodon/mastodon#22213]
### Pitch

This is valid webfinger acct `*@domain.tld` e.g. `*@johndoe.com`

According to [webfinger spec](https://www.rfc-editor.org/rfc/rfc7565#section-7) the userpart includes all ASCII 7 bit characters, including `*`:

> The userpart consists only of Unicode code points that conform to the PRECIS IdentifierClass specified in [[RFC7564](https://www.rfc-editor.org/rfc/rfc7564)].

[PRECIS IdentifierClass valid character](https://www.rfc-editor.org/rfc/rfc7564#section-4.2) sections says: 
> Code points in the range U+0021 through U+007E, i.e., the (printable) ASCII7 ("K") category defined under [Section 9.11](https://www.rfc-editor.org/rfc/rfc7564#section-9.11). These code points are "grandfathered" into PRECIS and thus are valid even if they would otherwise be disallowed according to the property-based rules specified in the next section.

However `*` is not valid username in Mastodon, thus it could be used to denote in webfinger that user handles part is just the domain.

I tried searching the issue, it must have come up before in some form, but couldn't find it.

### Motivation

People and domain owners generally would like to have recognizable user handle. Currently those who own e.g. domain `johndoe.com` have to make their handles cumbersome by adding their name again e.g. `johndoe@johndoe.com`.

This same applies to brands, e.g. `@apple@apple.com` looks odd, but `@apple.com` would suit them better.

This could be fixed if one could at directly `@firstnamelastname.tld`

**Comment by snarfed** [comment]
Yes! https://fed.brid.gy/ adds bidirectional ActivityPub to arbitrary web sites and blogs; we would happily use this. Right now we use `@domain.tld@domain.tld`, which works, but is obviously awkward.

**Comment by filipesaraiva** [comment]
I hope this can be integrated soon. But it is a change for ActivityPub protocol itself or just Mastodon? I think other ActivityPub apps don't handle it yet.

**Comment by trwnh** [comment]
> But it is a change for ActivityPub protocol itself or just Mastodon?

webfinger isn't strictly part of activitypub, so it would be on mastodon and i suppose anything that wants to be compatible with mastodon.

with regards to the specific proposal: i *think* there's some existing pattern / precedent for using `acct:domain.tld@domain.tld`, but really the bigger issue if you want to use a bare `domain.tld` is going to be the expectation that `preferredUsername` is set on the activitypub actor, and specifically that `${preferredUsername}@domain` must be fed back into webfinger to yield the exact same activitypub resource. so for example, if your preferred username is `apple` and your activitypub id is `https://apple.com` then mastodon will extract the username and the domain and verify that webfinger leads back to the same activitypub `id`.

tagging #17030 as it may make more sense to ditch the strict webfinger requirement and instead switch to a sort of contact/address-book model where you keep track of people by their `id` or url or petname or whatever

also for what it's worth, i don't think you need to specify an asterisk `*` in webfinger. insofar as we're still using webfinger (and ignoring additional requirements imposed by mastodon), you could just leave out the userpart. `acct:@apple.com` is a valid acct uri, and by webfinger, you would extract the host as `apple.com` and then perform `GET https://apple.com/.well-known/webfinger?resource=acct:@apple.com`.

**Comment by snarfed** [comment]
Hah, as a data point, https://fed.brid.gy/ 's webfinger already supports `acct@domain`: https://fed.brid.gy/.well-known/webfinger?resource=acct:@snarfed.org

**Comment by manton** [comment]
I would love to see this. Domain names as usernames is much more natural for single-user Mastodon instances and blogs.

**Comment by manton** [comment]
I like the Nostr solution to this using an underscore: `_@domain`. Details in [NIP-5](https://github.com/nostr-protocol/nips/blob/master/05.md). I feel like this could be more compatible with existing software than an asterisk `*`, but that's just a guess.

Behind the scenes in WebFinger and ActivityPub, `_@domain` would still be used. But in the UI, only `@domain` would be shown.

Would love to hear other thoughts on this.

**Comment by tesaguri** [comment]
> Hah, as a data point, https://fed.brid.gy/ 's webfinger already supports `acct@domain`: https://fed.brid.gy/.well-known/webfinger?resource=acct:@snarfed.org

Interesting, but IIUC that's in violation of [RFC 7565 Section 7.], which specifies that the `userpart` component of an `acct:` URI contains at least one character.

(And it seems that the WebFinger URI you showed returns `400` now.)

[RFC 7565 Section 7.]: <https://datatracker.ietf.org/doc/html/rfc7565#section-7>

**Comment by manton** [comment]
@snarfed Does it make sense for Bridgy to support the underscore like `acct:_@domain` so that it's technically compatible with RFC 7565? Would be great to have some consensus around this for apps to experiment with.

**Comment by snarfed** [comment]
> Does it make sense for Bridgy to support the underscore like `acct:_@domain` so that it's technically compatible with RFC 7565?

Hmm sure! Afaik Bridgy Fed is compatible right now; it officially uses `@domain@domain`, and now 400s `acct:@domain`, as mentioned in https://github.com/mastodon/mastodon/issues/22213#issuecomment-1741621572.

I'm definitely on board with Nostr style `_@domain` being a shorthand for just `domain`! Afaik most of that is just UI sugar in clients so that they hide the `_@` when they see an `_@domain`; backends like BF wouldn't do much other than maybe encourage people to use the `_` username on single-user domains. I'm happy to start doing that once we see some of the `_@` UI hiding in clients!

**Comment by tesaguri** [comment]
Related Fediverse Idea issue: [#23 - Domain names as handles - fediverse-ideas - Codeberg.org][fediverse-ideas-23].

I've raised a concern regarding the `@_` approach there. The gist is that, I don't think the underscore username should be used for representing the domain since Mastodon currently assigns it to unprivileged users.

[fediverse-ideas-23]: <https://codeberg.org/fediverse/fediverse-ideas/issues/23>

**Comment by pfefferle** [comment]
Why trying to mimic username@domain.tld at all? WebFinger also supports URLs, so why not simply translate: `@username@domain.tld` to `acct:username@domain.tld` and `@domain.tld` to `https://domain.tld`.

This way we do not have to use wildcards or catch-all usernames at all, we simply have to change the @-reply mapping.

**Comment by mbajur** [comment]
I'm working on a single-user kind of a web app (one user - one domain) and that would be amazing to be able to follow such users directly via the domain they are using. Hope this issue is gonna be worked on at some point 🤞

**Comment by manton** [comment]
Hello everyone! I'd like to revisit this. One additional question is how to handle the `preferredUsername` field in actor JSON. This is usually the `@username` part of `@username@example.com`. Would it break anything if `preferredUsername` was only an underscore?

Ideally I would agree with @pfefferle about passing simple URLs to WebFinger. But I wonder about having some fallbacks for ActivityPub implementations that aren't likely to be updated.

**Comment by trwnh** [comment]
mastodon should allow a single underscore, i think? https://swicg.github.io/activitypub-webfinger/#rules

although i would still just do preferredUsername = domain.example (resulting in `@domain.example@domain.example`) or preferredUsername = foo (resulting in `@foo@domain.example`). if you do an underscore that would get rendered as `@_` in text content which would get confusing without the domain name visible -- because mastodon typically hides the domainpart unless there is ambiguity between two mentions in the same status

the challenge with changing the @ mapping is that it's not clear whether the mention is of a domain or of a username. usernames are allowed to have dots in them.

similarly, preferredUsername = _ will be interpreted as a username that is an underscore, and it's not clear whether someone wants to be an underscore or their domain.
